### ISSUE TITLE
no audio from OS X aggregate audio device

### ISSUE THREAD

**Issue by holgafreak** [csound/csound#862]
Running FLOSSMANUAL vbap-examples using aggregate device gives no audio. Writing to a file works, and with Audition outputting to aggregate device the example moves sound ok.
Debugging shows, csound is waiting for semaphore in kernel forever. Interrupting with ctrl-c and then  continuing, a short (buffer length?) audio  is output, and so on.
Code stops on line 105 in CsoundLib64 audio.c
When running 4-ch example to 6-ch audio device, everything is ok.
I tried csound 6.09 from pkg and also latest git. I'm on OS X 10.11.6
-m

**Comment by vlazzarini** [comment]
We would need some more info:
1. How is the aggregate device configured: how many channels in/out, devices etc.
2. Do the separate original devices work OK with Csound?
3. What backend are you using: portaudio, coreaudio or jack? Did you try with 
different backends?
4. There is no audio.c in the Csound sources, maybe that file is part of the system libs?
5. The case where everything is OK is with the original device, not aggregate?
6. Can you run one or two channels to aggregate device?

Finally can you copy here the output you are getting off the terminal?

**Comment by holgafreak** [comment]
1. aggregate dev is 6-ch usb audio card + 2-ch builtin output, so 8-ch total
2. yes, both dev works as separate devs
3. both coreaudio and portaudio
4. sorry, I wrote earlier from my linux laptop,. now here's the lldb output:
(lldb) bt
* thread #1: tid = 0x16ee61, 0x00007fff909dc10a libsystem_kernel.dylib`__semwait_signal + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff909dc10a libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff80719d0f libsystem_c.dylib`nanosleep + 199
    frame #2: 0x00007fff80719c02 libsystem_c.dylib`usleep + 54
    frame #3: 0x000000010399857f librtauhal.dylib`rtplay_ + 239
    frame #4: 0x000000010004f846 CsoundLib64`spoutsf(csound=0x0000000102800000) + 646 at libsnd.c:105
    frame #5: 0x000000010000979d CsoundLib64`kperf_nodebug(csound=0x0000000102800000) + 2477 at csound.c:1692
    frame #6: 0x000000010000b21b CsoundLib64`csoundPerform(csound=0x0000000102800000) + 411 at csound.c:2143
    frame #7: 0x0000000100001add csound`main(argc=3, argv=0x00007fff5fbff6e8) + 781 at csound_main.c:322
    frame #8: 0x00007fff909c45ad libdyld.dylib`start + 1
    frame #9: 0x00007fff909c45ad libdyld.dylib`start + 1
-------
(lldb) f 4
frame #4: 0x000000010004f846 CsoundLib64`spoutsf(csound=0x0000000102800000) + 646 at libsnd.c:105
   102 	    if (!csound->libsndStatics.outbufrem) {
   103 	      if (csound->libsndStatics.osfopen) {
   104 	        csound->nrecs++;
-> 105 	        csound->audtran(csound, csound->libsndStatics.outbuf,
   106 	                        csound->libsndStatics.outbufsiz); /* Flush buffer */
   107 	        csound->libsndStatics.outbufp = (MYFLT*) csound->libsndStatics.outbuf;
   108 	      }
5. Adobe Audition plays the 8-ch vbap-example generated wav-file correctly, 4-ch vbap-example works with the 6 -ch usb card
6. No, same deadlock

terminal:
orch now loaded
audio buffered in 1024 sample-frame blocks
==========================================================
AuHAL Module: found 10 output device(s):
1: Built-in Output (2 channels) 
2: unnamed device (2 channels) 
3: Generic USB Audio Device    (2 channels) 
4: unnamed device (6 channels) 
5: Soundflower (2ch) (2 channels) 
6: Soundflower (64ch) (64 channels) 
7: NoMachine Audio Adapter (2 channels) 
8: NoMachine Microphone Adapter (2 channels) 
9: kuwatec Audio Loopback (2 channels) 
10: Aggregate Device (8 channels) 
selected output device: Aggregate Device 
***** AuHAL module: output device open with 1024 buffer frames
==========================================================
writing 8192 sample blks of 64-bit floats to dac10 
SECTION 1:
new alloc for instr 12:
diskin2: opened '/Users/mjkoskin/motari2.wav':
         48000 Hz, 2 channel(s), 763687 sample frames
^C
csound command: Interrupt
csoundPerform(): stopped.
inactive allocs returned to freespace
end of score.		   overall amps:  0.01672  0.02133  0.01672  0.02133  0.01672  0.02133  0.01672  0.02133
	   overall samples out of range:        0        0        0        0        0        0        0        0
0 errors in performance
Elapsed time at end of performance: real: 4.273s, CPU: 0.058s
AuHAL module: device closed
4 8192 sample blks of 64-bit floats written to dac10

**Comment by vlazzarini** [comment]
Thanks for this, we have something to investigate now.
Can you try jack, if possible?
Also is the behaviour exactly the same with portaudio and coreaudio?

**Comment by vlazzarini** [comment]
I had a look at your backtrace and it's just reporting normal operation (waiting on the circular buffer), so it does look like Csound is playing. However, it might be that the coreaudio callback is blocking perhaps. Let me think about this.

**Comment by vlazzarini** [comment]
Another thing to test: can you make an aggregate device consisting of only one of the two devices you are using (the two or the six-channel one) and see if that works.

**Comment by vlazzarini** [comment]
also another test: can you add a 
printk 0.2, time:k()

to your running instrument. I want to see when it stops. The amps in your printout seem to indicate that some audio was synthesised.

**Comment by holgafreak** [comment]
all the instruments are working after adding printk 0.2,times:k()
they play just normally and csound ends afterwards.
So a mutex or some other lock perhaps in a thread? But why on an aggregate device? I remember, that 8-ch aggregate device and an earlier version of csound, and probably an earlier version of osx, too, there were no problems.

**Comment by vlazzarini** [comment]
What do you mean by 'play normally'? They are playing but you don't get any sound?
If they are playing, I'd say there is no deadlock, but maybe something else preventing
the audio output.

If the audio device weren't pulling the audio, I'd say Csound would hang rather than run
continuously.

**Comment by holgafreak** [comment]
I meant that I get also audio, the instruments don't hang, no need to ctrl-c to end csound. The only needed thing was to add the printk, so there's something for csound other than just play. I also tried to write, that with this printk, some thread timings get correct. So, playing 8-ch audio with aggregate device, adding printk to the instruments, solves the no audio problem.

**Comment by vlazzarini** [comment]
could you try removing the printk and setting ksmps = 1 and see if it plays as well? Hard to diagnose from a distance.

**Comment by holgafreak** [comment]
it works setting ksmps=1

**Comment by vlazzarini** [comment]
interesting. It seems to be fixed by slow Csound down. Do you have the same behaviour on portaudio (rtpa) (I see the lldb backtrace was created using rtauhal)?

**Comment by holgafreak** [comment]
yes, using portaudio was the same, no audio output. I've never succeeded with jack getting audio even from 2-ch directly to builtin output, so I did not try jack.

**Comment by vlazzarini** [comment]
ok and the portaudio gets fixed in the same way? I am trying to pin down likely places for this.

**Comment by holgafreak** [comment]
yes, ksmps=1 and portaudio -> audio ok

**Comment by holgafreak** [comment]
setting sr=44100 and ksmps=10 works too, although devices are 48000. Does osx or csound resample by fly?

**Comment by holgafreak** [comment]
still one, using ksmps=1 or 10, 6144 sample blks are written, the original example had ksmps=32 and 1024 blks were written.

**Comment by holgafreak** [comment]
sorry, from the earlier tests nchnls were left to 6, not 8. Wit 8-ch either ksmps=1 or printk doesn't give any audio. I'm really sorry for the confusion. So we're back to square one. My bad.

**Comment by vlazzarini** [comment]
does printk print anything?

**Comment by holgafreak** [comment]
found the reason, in my original csd I had had -b 30 in the CsOptions, but the floss examples didn't have this. I added it back, and all 8-channels are singing. Should there be a warning, if buffers are not set? Now I remember, that I tried many, many settings to keep coreaudio happy. In some earlier osx version, coreaudio output a warning, and I finally settled to 30. Now if i set ksmps to 32, there are small pauses in the audio, setting ksmps=1 all work fine. (fingers crossed)
